### PR TITLE
gh-102812: unittest's addError/addFailure/addSubTest accept an exception instance in addition to (typ, val, tb) as paramter

### PR DESCRIPTION
**Pull request by iritkatriel**
* Issue: gh-102812

### CODE REVIEW

File: Lib/unittest/case.py [hunk 1]
```diff
@@ -91,12 +90,13 @@ def _addSkip(result, test_case, reason):
...
-        if issubclass(exc_info[0], test.failureException):
-            result.addFailure(test, exc_info)
+def _addError(result, test, exc):
+    assert not isinstance(exc, tuple)
+    if result is not None and exc is not None:
+        if isinstance(exc, test.failureException):
+            result.addFailure(test, exc)
         else:
-            result.addError(test, exc_info)
+            result.addError(test, exc)
```
Comment (iritkatriel) [inline Lib/unittest/case.py:100]: Maybe I should revert this change, in case there are use-defined subclasses of TestResult that expect the tuple?

File: Lib/unittest/case.py [hunk 2]
```diff
@@ -567,8 +567,8 @@ def _addUnexpectedSuccess(self, result):
             # otherwise the legacy result can choke.
             try:
                 raise _UnexpectedSuccess from None
-            except _UnexpectedSuccess:
-                result.addFailure(self, sys.exc_info())
+            except _UnexpectedSuccess as e:
+                result.addFailure(self, e)
```
Comment (iritkatriel) [inline Lib/unittest/case.py:572]: ditto.

File: Lib/unittest/suite.py [hunk 3]
```diff
@@ -179,9 +179,10 @@ def _handleClassSetUp(self, test, result):
                 if failed and doClassCleanups is not None:
                     doClassCleanups()
                     for exc_info in currentClass.tearDown_exceptions:
+                        exc = exc_info[1] if isinstance(exc_info, tuple) else exc_info
                         self._createClassOrModuleLevelException(
                                 result, exc_info[1], 'setUpClass', className,
-                                info=exc_info)
+                                err=exc)
```
Comment (iritkatriel) [inline Lib/unittest/suite.py:185]: ditto

File: Lib/unittest/case.py (hunk 1 above)
Comment (cjw296) [reply Lib/unittest/case.py:100]: Yeah, I think the folks still using unittest are often in interesting and subtle places (Twisted's trial springs to mind...) so I think we really want to make sure 100% backwards compatibility (even in internal APIs *sigh*) is preserved if at all possible.

I'm curious: where are you finding yourself calling `addError` and friends in your code?

Comment (iritkatriel) [reply Lib/unittest/case.py:100]: I'm not calling addError in my code. We're trying to remove the (typ, exc, tb) triplet from the language. The interpreter doesn't need it anymore, it only exists for backwards compatibility.

File: Doc/whatsnew/3.12.rst [hunk 4]
```diff
@@ -385,6 +385,15 @@ typing
   same name on a base class, as per :pep:`698`. (Contributed by Steven Troxler in
   :gh:`101564`.)
 
+unittest
+--------
+
+:meth:`~unittest.TestResult.addError`,
+:meth:`~unittest.TestResult.addFailure` and
+:meth:`~unittest.TestResult.addSubTest` now accept either an exception
+instance or a ``(typ, val, tb)`` tuple. (Contributed by Irit Katriel in
```
Comment (terryjreedy) [inline Doc/whatsnew/3.12.rst:755]: Make clear what is new while being clear that still only 1 is passed.
```suggestion
:meth:`~unittest.TestResult.addSubTest` now accept an exception instead of
a ``(typ, val, tb)`` tuple. (Contributed by Irit Katriel in
```
Do you want to say that an exception is in some sense preferred?

File: Lib/unittest/runner.py [hunk 5]
```diff
@@ -73,13 +73,14 @@ def _write_status(self, test, status):
 
     def addSubTest(self, test, subtest, err):
         if err is not None:
+            err = err[1] if isinstance(err, tuple) else err
             if self.showAll:
-                if issubclass(err[0], subtest.failureException):
+                if isinstance(err, subtest.failureException):
                     self._write_status(subtest, "FAIL")
                 else:
                     self._write_status(subtest, "ERROR")
```
Comment (terryjreedy) [inline Lib/unittest/runner.py:84]: Factor out minimal difference in the two branches depending on the condition.
```suggestion
                status = "FAIL" if isinstance(err, subtest.failureException) else "ERROR"
                self._write_status(subtest, status)
```
There might be a better word than 'status'.  Note: if the statement were not already being edited, this would be an unrelated change and I would not suggest it here.  But since you are ...

File: Lib/unittest/runner.py [hunk 6]
```diff
@@ -73,13 +73,14 @@ def _write_status(self, test, status):
...
             elif self.dots:
-                if issubclass(err[0], subtest.failureException):
+                if isinstance(err, subtest.failureException):
                     self.stream.write('F')
                 else:
                     self.stream.write('E')
```
Comment (terryjreedy) [inline Lib/unittest/runner.py:89]:
```suggestion
                dot = 'F' if isinstance(err, subtest.failureException) else 'E'
                self.stream.write(dot)
```

File: Lib/unittest/suite.py [hunk 7]
```diff
@@ -231,21 +232,22 @@ def _handleModuleFixture(self, test, result):
...
-                                        info=None):
+                                        err=None):
         error = _ErrorHolder(errorName)
         addSkip = getattr(result, 'addSkip', None)
         if addSkip is not None and isinstance(exception, case.SkipTest):
             addSkip(error, str(exception))
         else:
-            if not info:
-                result.addError(error, sys.exc_info())
+            if not err:
```
Comment (terryjreedy) [inline Lib/unittest/suite.py:246]: It is not clear why the assert it needed, but with it, a possible refactoring (which cannot be marked as a multi-line suggestion).
```
            if not err:
                err = sys.exc_info()[1]
            else:
                assert not isinstance(err, tuple)
            result.addError(error, err)
```

File: Misc/NEWS.d/next/Library/2023-03-18-13-09-35.gh-issue-102812.IcPYBs.rst [hunk 8]
```diff
@@ -0,0 +1,4 @@
+:meth:`~unittest.TestResult.addError`,
+:meth:`~unittest.TestResult.addFailure` and
+:meth:`~unittest.TestResult.addSubTest` now accept either an exception
+instance or a ``(typ, val, tb)`` tuple.
```
Comment (terryjreedy) [inline Misc/NEWS.d/next/Library/2023-03-18-13-09-35.gh-issue-102812.IcPYBs.rst:4]:
```suggestion
:meth:`~unittest.TestResult.addSubTest` now accept an exception instance
instead of a ``(typ, val, tb)`` tuple.
```

File: Doc/whatsnew/3.12.rst (hunk 4 above)
Comment (iritkatriel) [reply Doc/whatsnew/3.12.rst:755]: I think "instead of" may sound like the old format is no longer accepted. But agree I would emphasise what is new and what is old.

File: Lib/unittest/case.py (hunk 1 above)
Comment (iritkatriel) [reply Lib/unittest/case.py:100]: I made this a draft PR, it need more thought. If we can't make unittest use the new APIs then I'm not sure there is a point committing this. We would need a plan for how to migrate off the triplet. 

This is related to the PEP707 discussion, so we could wait to see what comes out of that.

File: Doc/library/unittest.rst [hunk 9]
```diff
@@ -2104,24 +2104,29 @@ Loading and running tests
...
-      traceback)``.
+      Called when the test case *test* raises an unexpected exception. *err* is
+      an exception instance or a tuple of the form returned by
+      :func:`sys.exc_info`: ``(type, value, traceback)``.
 
       The default implementation appends a tuple ``(test, formatted_err)`` to
       the instance's :attr:`errors` attribute, where *formatted_err* is a
       formatted traceback derived from *err*.
 
+      .. versionchanged:: 3.12
```
Comment (terryjreedy) [inline Doc/library/unittest.rst:2115]:
```suggestion
      .. versionchanged:: 3.13
```

File: Doc/library/unittest.rst [hunk 10]
```diff
@@ -2104,24 +2104,29 @@ Loading and running tests
...
-      the form returned by :func:`sys.exc_info`: ``(type, value, traceback)``.
+      Called when the test case *test* signals a failure. *err* is an exception
+      instance or a tuple of the form returned by :func:`sys.exc_info`:
+      ``(type, value, traceback)``.
 
       The default implementation appends a tuple ``(test, formatted_err)`` to
       the instance's :attr:`failures` attribute, where *formatted_err* is a
       formatted traceback derived from *err*.
 
+      .. versionchanged:: 3.12
```
Comment (terryjreedy) [inline Doc/library/unittest.rst:2128]:
```suggestion
      .. versionchanged:: 3.13
```

File: Doc/library/unittest.rst [hunk 11]
```diff
@@ -2165,15 +2170,19 @@ Loading and running tests
...
+      or a tuple of the form returned by :func:`sys.exc_info`:
+      ``(type, value, traceback)``.
 
       The default implementation does nothing when the outcome is a
       success, and records subtest failures as normal failures.
 
       .. versionadded:: 3.4
 
-   .. method:: addDuration(test, elapsed)
+      .. versionchanged:: 3.12
```
Comment (terryjreedy) [inline Doc/library/unittest.rst:2182]:
```suggestion
      .. versionchanged:: 3.13
```

**Review by terryjreedy — COMMENTED** [review]
My turn to ping ;-).  Too late for 3.12 I guess, hence new suggestion(s).  Please accept the suggestions you are sure of, and resolve when appropriate.  I fixed the merge conflict (easy).

**Comment by iritkatriel** [comment]
Thanks, @terryjreedy . I have some doubts about this PR because I think there might be user-defined subclasses of unittest that override these functions (essentially callbacks expecting (typ, exc, tb) triplets). I'm not sure how to move us on from that.

**Comment by iritkatriel** [comment]
Closing because I think this is more complicated than I thought. (See previous comment).
